src/renderers/shared/fiber/ReactFiberBeginWork.js

```diff
@@ -450,7 +457,9 @@ module.exports = function<T, P, I, TI, PI, C, CX, PL>(
       'An indeterminate component should never have mounted. This error is ' +
       'likely caused by a bug in React. Please file an issue.'
     );
-    var fn = workInProgress.type;
+    var fn = hasMockingBehavior
+      ? mockComponent(workInProgress, hostContext.getRootHostContainer())
```

Comment:
That way this would be the only place we'd ever have to branch on `hasMockingBehavior`